cooked-validators/tests/Cooked/AttackSpec/DoubleSat.hs

```diff
@@ -0,0 +1,212 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cooked.AttackSpec.DoubleSat (tests) where
+
+import Control.Arrow
+import Cooked.Attack
+import Cooked.AttackSpec.Util
+import Cooked.MockChain
+import Cooked.Tx.Constraints
+import Cooked.Tx.Constraints.Optics
+import Data.Default
+import qualified Ledger.Ada as L
+import qualified Ledger.Typed.Scripts as L
+import qualified Ledger.Value as L
+import qualified Plutus.Script.Utils.V1.Scripts as L
+import qualified Plutus.V1.Ledger.Contexts as L
+import qualified PlutusTx as Pl
+import qualified PlutusTx.Eq as Pl
+import qualified PlutusTx.Prelude as Pl
+import Test.Tasty
+import Test.Tasty.HUnit
+
+-- * Mock contracts for the double satisfaction attack
+
+-- Scenario: There are two validators, one of type A, one of type B. We want to
+-- add an input belonging to the B validator to a transaction that spends from
+-- the A validator.
+
+data ADatum = ADatum deriving (Show)
+
+instance Pl.Eq ADatum where
+  ADatum == ADatum = True
+
+Pl.makeLift ''ADatum
+Pl.unstableMakeIsData ''ADatum
+
+data ARedeemer = ARedeemer deriving (Show)
+
+instance Pl.Eq ARedeemer where
+  ARedeemer == ARedeemer = True
+
+Pl.makeLift ''ARedeemer
+Pl.unstableMakeIsData ''ARedeemer
+
+data AContract
+
+instance L.ValidatorTypes AContract where
+  type DatumType AContract = ADatum
+  type RedeemerType AContract = ARedeemer
+
+{-# INLINEABLE mkAValidator #-}
+mkAValidator :: ADatum -> ARedeemer -> L.ScriptContext -> Bool
+mkAValidator _ _ _ = True
+
+aValidator :: L.TypedValidator AContract
+aValidator =
+  L.mkTypedValidator @AContract
+    $$(Pl.compile [||mkAValidator||])
+    $$(Pl.compile [||wrap||])
+  where
+    wrap = L.mkUntypedValidator @ADatum @ARedeemer
+
+data BDatum = BDatum deriving (Show)
+
+instance Pl.Eq BDatum where
+  BDatum == BDatum = True
+
+Pl.makeLift ''BDatum
+Pl.unstableMakeIsData ''BDatum
+
+data BRedeemer = BRedeemer1 | BRedeemer2 deriving (Show)
+
+instance Pl.Eq BRedeemer where
+  BRedeemer1 == BRedeemer1 = True
+  BRedeemer2 == BRedeemer2 = True
+  _ == _ = False
+
+Pl.makeLift ''BRedeemer
+Pl.unstableMakeIsData ''BRedeemer
+
+data BContract
+
+instance L.ValidatorTypes BContract where
+  type DatumType BContract = BDatum
+  type RedeemerType BContract = BRedeemer
+
+{-# INLINEABLE mkBValidator #-}
+mkBValidator :: BDatum -> BRedeemer -> L.ScriptContext -> Bool
+mkBValidator _ _ _ = True
+
+bValidator :: L.TypedValidator BContract
```

Comment:
There's a chance that `aValidator` and `bValidator` are the same and have the same address, we should test for that.